File: Ferrum/Errors/WrappedError.fs
namespace Ferrum

open System
open System.Diagnostics


type WrappedError<'e>(error: 'e) =

    interface IError with
        member this.Message = $"{error}"
        member this.InnerError = ValueNone

    override this.ToString() =
        ErrorFormatter.Default.Format(this)

    interface IFormattable with
        member this.ToString(format, _formatProvider) =
            (ErrorFormatter.byFormat format).Format(this)


type WrappedTracedError<'e>(error: 'e, stackTrace: StackTrace) =
    inherit WrappedError<'e>(error)

    [<StackTraceHidden>]
    new(error: 'e) = WrappedTracedError(error, StackTrace(0, true))

    interface ITracedError with
        member this.StackTrace = stackTrace.ToString()
        member this.LocalStackTrace = stackTrace

File: Ferrum.Core/Errors/WrappedError.cs
using System.Diagnostics;
using Ferrum.Formatting;

namespace Ferrum.Errors;



public class WrappedError<T>(T error) : IError, IFormattable
{

    public string Message => Internal.ToStringOrDefault(error);

    public IError? InnerError => null;

    public string ToString(string? format, IFormatProvider? formatProvider)
    {
        return (format is not null ? ErrorFormatter.ByFormat(format) : ErrorFormatter.Default).Format(this);
    }

    public override string ToString()
    {
        return ErrorFormatter.Default.Format(this);
    }
}

public class WrappedTracedError<T>(T error, StackTrace stackTrace) : WrappedError<T>(error), ITracedError
{
    [StackTraceHidden]
    public WrappedTracedError(T error) :
        this(error, new StackTrace(0, true)) {}
    public string? StackTrace => stackTrace.ToString();
    public StackTrace? LocalStackTrace => stackTrace;
}
